nomad/node_endpoint.go

```diff
@@ -1247,6 +1285,30 @@ func (n *Node) batchUpdate(future *structs.BatchFuture, updates []*structs.Alloc
 			trimmedEvals = append(trimmedEvals, eval)
 			evalsByJobId[namespacedID] = struct{}{}
 		}
+
+		// Update orphaned allocs so that migrate = false
+		if eval.TriggeredBy == structs.EvalTriggerJobDeregister || eval.TriggeredBy == structs.EvalTriggerReconnect {
+			for _, update := range updates {
+				if update.JobID != eval.JobID {
+					continue
+				}
+				// Don't update the desired transition unless the alloc is reconnecting
+				// after a drain or purge.
+				if eval.TriggeredBy == structs.EvalTriggerReconnect &&
+					update.DesiredStatus != structs.AllocDesiredStatusStop {
+					continue
+				}
+				transitionReq := &structs.AllocUpdateDesiredTransitionRequest{
+					Allocs: map[string]*structs.DesiredTransition{update.ID: &update.DesiredTransition},
+					Evals:  []*structs.Evaluation{eval},
+				}
+				_, index, err = n.srv.raftApply(structs.AllocUpdateDesiredTransitionRequestType, transitionReq)
```

Comment:
The `index` here is getting throw away here which upsets the linter, but in this case the lint is revealing a design problem: this raft entry isn't atomic between updates, and not atomic with the `AllocUpdateRequest` below. We can end up being unable to serve part of the request. Is there any way to get this all into a single raft entry (for example, by moving the logic into `nomad/state/state_store.go`)? We have a few existing cases in the code base where we don't have atomic requests and we pretty much regret all of them. 😀 

(Note that this endpoint already has this problem with the Consul/Vault tokens but that can't be helped because the Consul/Vault tokens are external to Nomad.)

Reply:
Thanks for letting me know that is the preferred way. I'll pivot that way.